File: oopjs/vidu1.js
const prompt = require('prompt-sync')();

class Canbo {
    constructor(mcb = "", hoten = "", gender = "") {
        this._mcb = mcb;
        this._hoten = hoten;
        this._gender = gender;
    }

    inp() {
        this._mcb = prompt("Nhập MCB: ");
        this._hoten = prompt("Nhập Họ Tên: ");
        this._gender = prompt("Nhập Giới Tính: ");
    }

    display() {
        console.log(`MCB: ${this._mcb}`);
        console.log(`Họ tên: ${this._hoten}`);
        console.log(`Giới tính: ${this._gender}`);
    }

    getGender() {
        return this._gender;
    }
}

class Congnhan extends Canbo {
    constructor(mcb = "", hoten = "", gender = "", bac = 1, ngaylam = 0, day = 1, month = 1, year = 2000) {
        super(mcb, hoten, gender);
        this._bac = bac;
        this._ngaylam = ngaylam;
        this._day = day;
        this._month = month;
        this._year = year;
    }

    inp() {
        super.inp();
        this._bac = parseInt(prompt("Nhập Bậc: "));
        this._ngaylam = parseInt(prompt("Nhập Ngày làm: "));
        this._day = parseInt(prompt("Nhập Ngày hợp đồng: "));
        this._month = parseInt(prompt("Nhập Tháng hợp đồng: "));
        this._year = parseInt(prompt("Nhập Năm hợp đồng: "));
    }

    display() {
        super.display();
        console.log(`Bậc: ${this._bac}`);
        console.log(`Ngày làm: ${this._ngaylam}`);
        console.log(`Hợp đồng: ${this._day}/${this._month}/${this._year}`);
        console.log(`Lương: ${this.luong()} VND`);
    }

    luong() {
        let cong = this._bac === 1 ? 300000 : this._bac === 2 ? 350000 : 400000;
        return this._ngaylam * cong;
    }
}

function main() {
    let n = parseInt(prompt("Nhập số lượng công nhân: "));
    if (n < 1 || n > 50) {
        console.log("Không hợp lệ");
        return;
    }

    let mang = [];
    for (let i = 0; i < n; i++) {
        let cn = new Congnhan();
        cn.inp();
        mang.push(cn);
    }

    mang.forEach(cn => {
        cn.display();
        console.log("<------------>");
    });

    let maxluong = Math.max(...mang.filter(cn => cn.getGender().toLowerCase() === "nam").map(cn => cn.luong()), 0);

    mang.forEach(cn => {
        if (cn.getGender().toLowerCase() === "nam" && cn.luong() === maxluong) {
            cn.display();
            console.log("-----------------------");
        }
    });
}

main();

File: oop.c++/vidu1.cpp
#include <bits/stdc++.h>
using namespace std;

class Canbo {
private:
    string macb;
    string hoten;
    string gender;

public:
    Canbo(string macb = "", string hoten = "", string gender = "") : macb(macb), hoten(hoten), gender(gender) {}

    string getmacb() const {
        return macb;
    }

    string gethoten() const {
        return hoten;
    }

    string getgender() const {
        return gender;
    }

    virtual void nhap() {
        getline(cin, macb);
        getline(cin, hoten);
        getline(cin, gender);
    }

    virtual void in() {
        cout << "MaCB: " << macb << endl;
        cout << "Ho ten: " << hoten << endl;
        cout << "Gender: " << gender << endl;
    }
};

class Congnhan : public Canbo {
private:
    int bac;
    int ngaylam;
    int day, month, year;

public:
    Congnhan(string macb = "",
             string hoten = "",
             string gender = "",
             int bac = 1,
             int ngaylam = 0,
             int day = 1,
             int month = 1,
             int year = 2000)
        : Canbo(macb, hoten, gender), bac(bac), ngaylam(ngaylam), day(day), month(month), year(year) {}

    int getbac() const {
        return bac;
    }

    int getngaylam() const {
        return ngaylam;
    }

    int getday() const {
        return day;
    }

    int getmonth() const {
        return month;
    }

    int getyear() const {
        return year;
    }

    void nhap() override {
        Canbo::nhap();
        cin >> bac;
        cin >> ngaylam;
        cin >> day >> month >> year;
        cin.ignore();
    }

    void in() override {
        Canbo::in();
        cout << "Bac: " << bac << endl;
        cout << "Ngay lam: " << ngaylam << endl;
        cout << "Ngay ki hop dong: " << day << "/" << month << "/" << year << endl;
        cout << "Luong: " << Luong() << endl;
    }

    int Luong() const {
        int cong;
        if (bac == 1) {
            cong = 300000;
        } else if (bac == 2) {
            cong = 350000;
        } else {
            cong = 400000;
        }
        return ngaylam * cong;
    }

    bool operator>(const Congnhan& other) const {
        if (year != other.year) {
            return year > other.year;
        }
        if (month != other.month) {
            return month > other.month;
        }
        return day > other.day;
    }
};

int main() {
 /*  Congnhan s1;
    s1.nhap();
    s1.in();
    return 0;
    */
    int n;
    cin>>n;
    Congnhan danhsach[50];
    for(int i=0;i<n;i++){
    	danhsach[i].nhap();
	}
	for(int i=0;i<n-1;i++){
		for(int j=i+1;j<n;j++){
			if(danhsach[j]>danhsach[i]){
				swap(danhsach[i],danhsach[j]);
			}
		}
	}
	for (int i = 0; i < n; ++i) { 
	danhsach[i].in(); 
	cout << "-----------------------\n";
    }

	int maxluong = 0;
	for (int i = 0; i < n; ++i) { 
	if (danhsach[i].getgender() == "Nam") { 
	maxluong = max(maxluong, danhsach[i].Luong()); 
		} 
	}
	for(int i = 0; i < n; ++i) { 
	if (danhsach[i].getgender() == "Nam" && danhsach[i].Luong() == maxluong) {
	 	danhsach[i].in(); 
	 	cout << "-----------------------\n"; 
		 } 
	}
	return 0;
}
